server/src/main/java/org/elasticsearch/cluster/routing/allocation/allocator/ClusterBalanceStats.java

```diff
@@ -196,23 +229,37 @@ private static NodeBalanceStats createFrom(
                 forecastWriteLoad += writeLoadForecaster.getForecastedWriteLoad(indexMetadata).orElse(0.0);
                 forecastShardSize += indexMetadata.getForecastedShardSizeInBytes().orElse(shardSize);
                 actualShardSize += shardSize;
+                if (isDesiredShardAllocation(shardRouting, desiredBalance) == false) {
+                    undesired++;
+                }
             }
 
             return new NodeBalanceStats(
                 routingNode.nodeId(),
                 routingNode.node().getRoles().stream().map(DiscoveryNodeRole::roleName).toList(),
                 routingNode.size(),
+                undesired,
                 forecastWriteLoad,
                 forecastShardSize,
                 actualShardSize
             );
         }
 
+        private static boolean isDesiredShardAllocation(ShardRouting shardRouting, DesiredBalance desiredBalance) {
+            if (shardRouting.relocating()) {
```

Comment:
Couldn't it be the case that the shard is being relocated to an undesired node? can't we check that through the `ShardRouting`?

Reply:
>Couldn't it be the case that the shard is being relocated to an undesired node?

It could be when the shard started relocating to the node-1 and balance changed so that node-2 become desired.

>can't we check that through the ShardRouting?

My understanding is that during iteration we would have both source shard (with relocating status) and target shard (with initializing status). The desired check is going to be performed against initializing target shard.